hello circleci

asippel89 opened the pull request:
@jdrake @msmathers cc @ns-ckao This integrates CircleCI for testing and publishing. I followed [this guide](https://circleci.com/blog/continuously-deploying-python-packages-to-pypi-with-circleci/) for working out the deploy job exclusion logic. The way it's set up now, on pull requests this will execute the tests. According to that blog post, the deploy portion should only be triggered when there's a new git tag created, which happens when you create a new release in GitHub (I'll need to test this once it's merged). 

Take a look and let me know your thoughts on that logic for publishing to Pypi and also happy to respond to any feedback on the config itself.

jdrake commented on .circleci/config.yml:
```diff
@@ -0,0 +1,46 @@
+
+commands:
+  install-deps:
+    description: Install dependencies using pre-commit
```
```suggestion
    description: Install dependencies and initialize pre-commit hooks
```

jdrake commented on .circleci/config.yml:
```diff
@@ -0,0 +1,46 @@
+          name: Install dependencies
+
+jobs:
+  build:
```
This is really "test" since `tox` tests things. But I would just add the `run: tox` step to the deploy steps so deploy is one job. Also maybe call the job "publish" since "deploy" is more infrastructure-y.

jdrake commented on .circleci/config.yml:
```diff
@@ -0,0 +1,46 @@
+            FLIT_USERNAME=$PYPI_USERNAME FLIT_PASSWORD=$PYPI_PASSWORD flit publish
+          name: Publish to Pypi
+
+workflows:
```
Also add a PR workflow with a single job that runs pre-commit hooks, runs tox (tests), and builds with `flit`:

For the step that does the pre-commit hooks, here is talos's:
```yaml
- run:
    name: Run commit hooks
    command: |
      pre-commit run \
        --source "origin/${TARGET_BRANCH}" \
        --origin "origin/${CIRCLE_BRANCH}" \
        --show-diff-on-failure
```
CIRCLE_BRANCH is a default env var. You can set the TARGET_BRANCH env var by hitting the github api. See https://github.com/NarrativeScience/talos/blob/master/.circleci/commands/get-target-branch.yml#L7

ns-ckao replied:
you can also use this orb! https://circleci.com/orbs/registry/orb/narrativescience/ghpr

jdrake commented on .circleci/config.yml:
```diff
@@ -0,0 +1,46 @@
+      - install-deps
+      - run:
+          command: |
+            FLIT_USERNAME=$PYPI_USERNAME FLIT_PASSWORD=$PYPI_PASSWORD flit publish
```
The FLIT_USERNAME and FLIT_PASSWORD should be stored in the CircleCI settings as environment variables. No need to pass them explicitly here. 

Also you need to `flit build` before `flit publish`

asippel89 commented:
@jdrake @msmathers @ns-ckao OK I think I addressed the comments (and am using the NS github pull request orb!). Let me know if there are other comments. I'd like to try out creating a release and publishing next.

jdrake commented on .circleci/config.yml:
```diff
@@ -0,0 +1,52 @@
+    steps:
+      - ghpr/build-prospective-branch
+      - install-deps
+      - run: pre-commit run --show-diff-on-failure
```
I think by default this will only run tests against the previous commit (not the base branch) which is why we were providing the branch options. @ns-ckao correct me if im wrong though

asippel89 replied:
Ah got it, I assumed it would use the prospective branch I build in the first step. Will update

jdrake commented on .circleci/config.yml:
```diff
@@ -0,0 +1,52 @@
+      - test:
+          filters:
+            tags:
+              only: /.*/
```
Does this tag filter do what we want? Won't this run on merge to master too?

asippel89 replied:
Ah, yeah good find. I think it was needed before when they were one workflow but it shouldn't be necessary anymore.

asippel89 commented:
OK, things seem to be working as expected now!
